File: otto-android/app/src/main/java/to/ottomot/driftd/core/notify/TimeZoneSync.kt
package to.ottomot.driftd.core.notify

import android.content.Context
import to.ottomot.driftd.AppContainer
import to.ottomot.driftd.core.network.dto.PatchMeTimeZoneRequestDto
import java.time.ZoneId

/**
 * Reports the device IANA time zone when it changes (push register + foreground).
 * Mirrors iOS [TimeZoneSync].
 */
object TimeZoneSync {
    private const val PREFS_NAME = "otto_time_zone_sync"
    private const val KEY_LAST_REPORTED = "otto.lastReportedTimeZone"

    fun systemIanaId(): String = ZoneId.systemDefault().id

    suspend fun syncIfNeeded(container: AppContainer) {
        val authToken = container.sessionRepository.authTokenState.value?.trim()
        if (authToken.isNullOrEmpty()) return

        val current = systemIanaId()
        if (current.isEmpty()) return

        val prefs = container.application.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val cached = prefs.getString(KEY_LAST_REPORTED, null)
        if (cached == current) return

        container.dataRepository.patchMeTimeZone(PatchMeTimeZoneRequestDto(current)).fold(
            onSuccess = {
                prefs.edit().putString(KEY_LAST_REPORTED, current).apply()
            },
            onFailure = { /* retry next foreground */ },
        )
    }

    fun primeCacheFromServerTimeZone(context: Context, serverTimeZone: String?) {
        val tz = serverTimeZone?.trim()?.takeIf { it.isNotEmpty() } ?: return
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        if (!prefs.contains(KEY_LAST_REPORTED)) {
            prefs.edit().putString(KEY_LAST_REPORTED, tz).apply()
        }
    }
}

File: otto-mobile/otto-mobile/TimeZoneSync.swift
import Foundation

/// Reports the device IANA time zone to the server when it changes (push register + foreground).
enum TimeZoneSync {
    private static let lastReportedKey = "otto.lastReportedTimeZone"

    static var systemIANAIdentifier: String {
        TimeZone.current.identifier
    }

    static func syncIfNeeded(isAuthenticated: Bool) async {
        guard isAuthenticated else { return }
        let current = systemIANAIdentifier
        guard !current.isEmpty else { return }
        let cached = UserDefaults.standard.string(forKey: lastReportedKey)
        guard cached != current else { return }
        do {
            try await APIClient.shared.patchMeTimeZone(timeZone: current)
            UserDefaults.standard.set(current, forKey: lastReportedKey)
        } catch {
            // Retry on next foreground; do not block UI.
        }
    }

    /// Call after auth/me when server has a different TZ than device (reinstall).
    static func primeCacheFromServerTimeZone(_ serverTimeZone: String?) {
        guard let serverTimeZone, !serverTimeZone.isEmpty else { return }
        if UserDefaults.standard.string(forKey: lastReportedKey) == nil {
            UserDefaults.standard.set(serverTimeZone, forKey: lastReportedKey)
        }
    }
}
